### PR TITLE
Feature/metadata ont

### PR DESCRIPTION
**Pull request by sven1103**

### CODE REVIEW

File: src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeMeasurement.groovy [hunk 1]
```diff
@@ -48,6 +61,16 @@ final class OxfordNanoporeMeasurement {
         return new OxfordNanoporeMeasurement(name, path, children, metadata)
     }
 
+    private static void validateMetaData(Map metadata) throws IllegalArgumentException {
+        try {
+            MetaData.validateMetadata(metadata)
+        } catch (ValidationException e) {
+            // Aggregate the causing exceptions
+            def causes = e.getCausingExceptions().collect{ it.message  }.join("\n")
+            throw new IllegalArgumentException("The Nanopore metadata could not be collected.\nReason:\n$causes",)
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeMeasurement.groovy:70]: Ah very nice!

**Comment by sven1103** [comment]
[Create JSON schema for nanopore metadata](https://trello.com/c/lgWkr3QS/100-create-json-schema-for-nanopore-metadata)

File: src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeMeasurement.groovy [hunk 2]
```diff
@@ -59,12 +82,27 @@ final class OxfordNanoporeMeasurement {
...
     private void readMetaData(Map<String, String> metadata) {
         this.metadata[METADATA_FIELD.ASIC_TEMP] = metadata["asic_temp"]
-        this.metadata[METADATA_FIELD.INSTRUMENT] = metadata["instrument"]
-        this.metadata[METADATA_FIELD.FLOWCELL_ID] = metadata["flowcell_id"]
-        this.metadata[METADATA_FIELD.FLOWCELL_POSITION] = metadata["flowcell_position"]
-        this.metadata[METADATA_FIELD.FLOWCELL_TYPE] = metadata["flowcell_type"]
+        this.metadata[METADATA_FIELD.DEVICE_TYPE] = metadata["device_type"]
+        this.metadata[METADATA_FIELD.FLOWCELL_ID] = metadata["flow_cell_id"]
+        this.metadata[METADATA_FIELD.FLOWCELL_POSITION] = metadata["flow_cell_position"]
+        this.metadata[METADATA_FIELD.FLOWCELL_TYPE] = metadata["flow_cell_product_code"]
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeMeasurement.groovy:88]: Just looked it up. Good catch splitting the `flowcell` into `flow` and `cell`

**Comment by sven1103** [comment]
[Validate nanopore metadata against the JSON schema in the `OxfordNanoporeMeasurement` class](https://trello.com/c/k5MY6NB4/101-validate-nanopore-metadata-against-the-json-schema-in-the-oxfordnanoporemeasurement-class)

File: src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeMeasurement.groovy [hunk 3]
```diff
@@ -151,20 +189,36 @@ final class OxfordNanoporeMeasurement {
...
+        return metadata.get(METADATA_FIELD.GUPPY_VERSION)
+    }
+
+    /**
+     * Provides access to the library kit used.
+     * @return
+     */
+    String getLibraryKit() {
+        return metadata.get(METADATA_FIELD.LIBRARY_PREPARATION_KIT)
+    }
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeMeasurement.groovy:206]:
```suggestion
    String getLibraryPreparationKit() {
        return metadata.get(METADATA_FIELD.LIBRARY_PREPARATION_KIT)
    }
```
To be consistent with the naming of Getters in relation to the fields.

File: src/test/groovy/life/qbic/datamodel/datasets/datastructure/OxfordNanoporeMeasurementSpec.groovy [hunk 4]
```diff
@@ -70,13 +75,15 @@ class OxfordNanoporeMeasurementSpec extends Specification {
 
         when:
         def result = measurement.getRawDataPerSample(mockedExperiment)
+        def libraryKit = measurement.getLibraryKit()
```
Comment (KochTobi) [inline src/test/groovy/life/qbic/datamodel/datasets/datastructure/OxfordNanoporeMeasurementSpec.groovy:78]:
```suggestion
        def libraryKit = measurement.getLibraryPreparationKit()
```

File: src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeMeasurement.groovy [hunk 5]
```diff
@@ -112,19 +150,19 @@ final class OxfordNanoporeMeasurement {
     }
 
     /**
-     * Provides access to the experiment start date.
+     * Provides access to the asic temperature.
      * @return
      */
-    String getStartDate() {
-        return metadata.get(METADATA_FIELD.START_DATE)
+    String getAsicTemp() {
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeMeasurement.groovy:156]: Good idea, sorting the getters alphabetically.

**Review by KochTobi — CHANGES_REQUESTED** [review]
One getter method breaks the naming schema.

**Comment by KochTobi** [comment]
DocStrings would also be nice ;)

File: src/main/resources/schemas/ont-metadata.schema.json [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    },
+    "guppy_version":  {
+      "type": "string",
+      "minLength": 1
+    },
+    "hostname": {
+      "type": "string",
+      "minLength": 1
+    },
+    "protocol": {
```
Comment (Steffengreiner) [inline src/main/resources/schemas/ont-metadata.schema.json:36]: Is there a reason, that the "protocol" property doesn't have a "minLength" attribute? Can there be cases where no protocol is needed/specified?

**Comment by sven1103** [comment]
where did I miss some?

File: src/test/resources/valid-example.json [hunk 7]
```diff
@@ -5,10 +5,15 @@
...
+                "device_type" : "promethion",
+                "flow_cell_id": "PAE26306",
+                "flow_cell_product_code": "FLO-PRO002",
+                "flow_cell_position": "2-A3-D3",
+                "guppy_version":  "3.2.8+bd67289",
+                "hostname": "PCT0094",
+                "protocol": "sequencing/sequencing_PRO002_DNA:FLO-PRO002:SQK-LSK109:True",
+                "started": "2020-02-11T15:52:10.465982+01:00"
             },
             "path": "file:///Users/kocht/SoftwareDevelopment/libs/core-utils-lib/src/test/resources/dummyFileSystem/nanopore-instrument-output/validates/QABCD001AB_E12A345a01_PAE12345/20200122_1217_1-A1-B1-PAE12345_1234567a/",
```
Comment (Steffengreiner) [inline src/test/resources/valid-example.json:18]: Would it be possible to replace the absolute filePath with a relative filePath starting from src?

**Comment by KochTobi** [comment]
In all the private methods. Although those are not used outside the class documentation would make sense.

**Review by KochTobi — APPROVED** [review]
Looks good now.

File: src/main/resources/schemas/ont-metadata.schema.json (hunk 6 above)
Comment (sven1103) [reply src/main/resources/schemas/ont-metadata.schema.json:36]: well spotted! I will add it in a new PR

Comment (sven1103) [reply src/main/resources/schemas/ont-metadata.schema.json:36]: ah, I defined a pattern:

"SQK-.*(?=:)"

that is why the minLength property is not needed here

Comment (Steffengreiner) [reply src/main/resources/schemas/ont-metadata.schema.json:36]: Ah, that's a good point!

**Pull request merged**